rust/actyx/ax-aql/src/language/parse_utils.rs

```diff
@@ -113,3 +117,77 @@ impl<'a> Ext<'a> for P<'a> {
         Ok(self.as_str().parse::<T>()?)
     }
 }
+
+pub trait Spanned {
+    type Ret;
+    fn spanned(self, span: pest::Span) -> Self::Ret;
+}
+
+impl<T, E: ToString> Spanned for Result<T, E> {
+    type Ret = Result<T, pest::error::Error<Rule>>;
+    fn spanned(self, span: pest::Span) -> Self::Ret {
+        self.map_err(|e| {
+            let e = ErrorVariant::CustomError { message: e.to_string() };
+            pest::error::Error::new_from_span(e, span)
+        })
+    }
+}
+
+#[derive(Debug, Clone, Copy, Eq)]
+pub struct Span<'a, T> {
+    span: pest::Span<'a>,
+    value: T,
+}
+
+impl<'a, T: PartialEq> PartialEq for Span<'a, T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.value == other.value
+    }
+}
+
+impl<'a, T: Ord> Ord for Span<'a, T> {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.value.cmp(&other.value)
```

Comment:
What is being ordered by this? Shouldn't the `Ord` implementation take into account the span itself rather than just the value?

Reply:
I thought the same, until I saw where it is used: ordering expression trees in the computation of AGGREGATE clauses to find duplicates (and compute only once). This means that including the spans defeats the purpose.

This `Span` is just a container for extra information that isn’t supposed to influence the rest of the program. A more practical problem was the roundtrip test, in which it is … inconvenient to generate the correct spans together with the syntax tree, so that the comparison in the end returns `true`.